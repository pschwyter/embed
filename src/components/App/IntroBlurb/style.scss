@import "~style/utilities/colors";
@import "~style/utilities/variables";

$shadow-color: rgba(2, 14, 29, .08);
$intro-shadow: drop-shadow(0 0 2px $shadow-color);

.ada-embed-intro-blurb {
  position: fixed;
  right: 76px;
  bottom: 24px;
  opacity: 0;
  max-width: 224px;
  z-index: 10000;

  &--show {
    animation-name: messageBlowup;
    animation-duration: 1s;
    animation-fill-mode: forwards;
  }

  &--hide {
    animation-name: messageBlowdown;
    animation-duration: .5s;
    animation-fill-mode: forwards;
  }

  &--large {
    right: 88px;
  }

  &--x-large {
    right: 104px;
  }

  &:hover &__dismiss-button,
  &__dismiss-button--mobile-show {
    opacity: 1;
    pointer-events: all;
  }

  &__dismiss-button {
    position: absolute;
    top: -8px;
    transition: opacity 180ms;
    opacity: 0;
    border: 0;
    border-radius: 50%;
    background-image: url("https://static.ada.support/Clear.svg");
    background-repeat: no-repeat;
    background-position: center;
    background-size: 18px;
    cursor: pointer;
    width: 16px;
    height: 16px;
    filter: $intro-shadow;
    pointer-events: none;

    &:focus {
      opacity: 1;
      pointer-events: all;
    }
  }

  &__message {
    margin-bottom: 0;
    border: 0;
    border-radius: 12px;
    box-shadow: 0 2px 9px 0 rgba(2, 11, 41, 0.08), 0 0 0 1px rgba(4, 33, 66, 0.02);
    background-color: $color-white;
    cursor: pointer;
    padding: 8px 12px;
    width: auto;
    max-width: 224px;
    height: auto;
    max-height: 100px;
    overflow: hidden;
    font-size: 16px;
    resize: none;
    word-break: break-word;
  }
}

@keyframes messageBlowup {
  0% {
    transform: scale3d(0, 0, 0);
    opacity: 0;
  }

  60% {
    transform: scale3d(1.1, 1.1, 1.1);
    animation-timing-function: ease-in-out;
    opacity: 1;
  }

  100% {
    transform: scale3d(1, 1, 1);
    opacity: 1;
  }
}

@keyframes messageBlowdown {
  0% {
    transform: scale3d(1, 1, 1);
    opacity: 1;
  }

  25% {
    transform: scale3d(1, 1, 1);
    opacity: 1;
  }

  50% {
    transform: scale3d(1.1, 1.1, 1.1);
    animation-timing-function: ease-in-out;
    opacity: 1;
  }

  100% {
    display: none;
    transform: scale3d(0, 0, 0);
    opacity: 1;
  }
}
