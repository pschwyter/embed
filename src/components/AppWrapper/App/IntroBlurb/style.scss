@import "~style/utilities/colors";
@import "~style/utilities/variables";

$shadow-color: rgba(2, 14, 29, .08);
$intro-shadow: drop-shadow(0 0 2px $shadow-color);

.ada-embed-intro-blurb {
  position: fixed;
  opacity: 0;
  max-width: 224px;
  z-index: 10000;
  visibility: hidden;

  &--not-draggable {
    position: fixed;
    right: 76px;
    bottom: 24px;
  }

  &--show {
    animation: messageBlowup 1s forwards;
  }

  &--hide {
    animation: messageBlowdown .5s forwards;
  }

  &--large {
    right: 88px;
  }

  &--x-large {
    right: 104px;
  }

  &__dismiss-button {
    position: absolute;
    top: -8px;
    transition: opacity 180ms;
    opacity: 0;
    border: 0;
    border-radius: 50%;
    background-image: url("https://static.ada.support/Clear.svg");
    background-repeat: no-repeat;
    background-position: center;
    background-size: 18px;
    cursor: pointer;
    width: 16px;
    height: 16px;
    filter: $intro-shadow;
    pointer-events: none;
    background-color: transparent;

    &--mobile-show {
      top: -24px;
    }

    &:focus {
      opacity: 1;
      pointer-events: all;
    }
  }

  &:hover &__dismiss-button,
  &__dismiss-button--mobile-show {
    opacity: 1 !important;
    pointer-events: all;
    margin: 0 13px;
    left: -13px;
  }

  &__dismiss-button--mobile-show ~ &__message {
    margin: 0;
  }

  &__message {
    margin-bottom: 0;
    border: 0;
    border-radius: 12px;
    box-shadow: 0 2px 9px 0 rgba(2, 11, 41, 0.08), 0 0 0 1px rgba(4, 33, 66, 0.02);
    background-color: $color-white;
    cursor: pointer;
    padding: 8px 12px;
    width: auto;
    max-width: 224px;
    height: auto;
    max-height: 100px;
    overflow: hidden;
    font-size: 16px;
    resize: none;
    word-break: break-word;
    text-align: left;
    min-width: 44px;
  }
}

@keyframes messageBlowup {
  0% {
    transform: scale3d(0, 0, 0);
    visibility: visible;
    opacity: 0;
  }

  60% {
    transform: scale3d(1.1, 1.1, 1.1);
    visibility: visible;
    opacity: 1;
  }

  100% {
    transform: scale3d(1, 1, 1);
    visibility: visible;
    opacity: 1;
  }
}

@keyframes messageBlowdown {
  0% {
    visibility: visible;
    transform: scale3d(1, 1, 1);
    opacity: 1;
  }

  25% {
    visibility: visible;
    transform: scale3d(1, 1, 1);
    opacity: 1;
  }

  50% {
    visibility: visible;
    transform: scale3d(1.1, 1.1, 1.1);
    opacity: 1;
  }

  100% {
    visibility: hidden;
    transform: scale3d(0, 0, 0);
    opacity: 1;
  }
}
