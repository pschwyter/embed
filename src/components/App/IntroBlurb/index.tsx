import classnames from "classnames";
import Client from "models/Client";
import { Component, h } from "preact";
import "./style.scss";

const INTRO_BLURB_SPACING = 32;

interface InterfaceIntroBlurb {
  client: Client,
  isInMobile: boolean,
  toggleChat(): void
}

interface InterfaceState {
  animateIntroIn: boolean,
  animateIntroOut: boolean
}

export default class IntroBlurb extends Component<InterfaceIntroBlurb, InterfaceState> {
  constructor(props: InterfaceIntroBlurb) {
    super(props);

    this.state = {
      animateIntroIn: false,
      animateIntroOut: false
    };

    this.dismissIntro = this.dismissIntro.bind(this);
    this.handleKeyPress = this.handleKeyPress.bind(this);
    this.handleOpenChat = this.handleOpenChat.bind(this);
  }

  componentDidMount() {
    const { client } = this.props;

    this.setState({
      animateIntroIn: true
    });

    setTimeout(() => {
      this.dismissIntro();
    }, (client.intro.duration + client.intro.delay) * 1000);
  }

  /**
   * Set animateIntroOut state to trigger animation
   */
  dismissIntro() {
    this.setState({
      animateIntroOut: true
    });
  }

  /**
   * Open chat and dismiss the intro
   */
  handleOpenChat() {
    const { toggleChat } = this.props;

    this.dismissIntro();
    toggleChat();
  }

  /**
   * Key press event handler
   */
  handleKeyPress(event: KeyboardEvent) {
    if (event.keyCode === 13) {
      this.handleOpenChat();
    }
  }

  render(props: InterfaceIntroBlurb) {
    const { client, isInMobile } = props;
    const { animateIntroIn, animateIntroOut } = this.state;

    return (
      <div
        className={
          classnames(
            "ada-embed-intro-blurb",
            {
              "ada-embed-intro-blurb--hide": animateIntroOut,
              "ada-embed-intro-blurb--show": animateIntroIn
            }
          )
        }
        style={{ right: client.chatButton.size + INTRO_BLURB_SPACING }}
      >
        <button
          className={classnames(
            "ada-embed-intro-blurb__dismiss-button", {
              "ada-embed-intro-blurb__dismiss-button--mobile-show": isInMobile
            }
          )}
          title="Dismiss Intro"
          onClick={this.dismissIntro}
        />
        <p
          className="ada-embed-intro-blurb__message"
          aria-live="assertive"
          role="alert"
          onClick={this.handleOpenChat}
          onKeyPress={this.handleKeyPress}
          tabIndex={0}
        >
          {client.intro.body}
        </p>
      </div>
    );
  }
}
