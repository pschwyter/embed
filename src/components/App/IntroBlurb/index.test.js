/* eslint-env node, jest */

import "jsdom-global/register";
import { shallow } from "preact-render-spy";
import IntroBlurb from "./index";

function setup(propsOverride = {}) {
  const defaultsProps = {
    client: {
      tint: "#000",
      intro: {
        duration: 1000,
        delay: 1000
      },
      chatButton: {
        size: 72,
        icon_path: "https://static.ada.support/images/286ca5d2-5311-467a-a5f5-5051ad710db4.svg",
        icon_type: "default",
        background_color: "#3ED1FF"
      }
    },
    toggleChat: jest.fn()
  };

  const props = Object.assign(defaultsProps, propsOverride);
  const PRSWrapper = shallow(<IntroBlurb {...props} />);

  return {
    props,
    PRSWrapper
  };
}

describe("<IntroBlurb />", () => {
  it("should match snapshot", done => {
    const { PRSWrapper } = setup();

    setTimeout(() => {
      expect(PRSWrapper).toMatchSnapshot();
      done();
    }, 0);
  });

  it("should have a tabIndex of 0 and aria-live assertive on the blurb message", () => {
    const { PRSWrapper, props } = setup();
    const messageElement = PRSWrapper.find(".ada-embed-intro-blurb__message");

    expect(messageElement.attr("tabIndex")).toEqual(0);
    expect(messageElement.attr("aria-live")).toEqual("assertive");
  });

  it("should open chat when clicking the blurb message", () => {
    const { PRSWrapper, props } = setup();
    const messageElement = PRSWrapper.find(".ada-embed-intro-blurb__message");

    messageElement.simulate("click");
    expect(props.toggleChat.mock.calls.length).toBe(1);
  });
});
