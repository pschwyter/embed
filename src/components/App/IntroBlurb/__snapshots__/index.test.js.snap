// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IntroBlurb /> should match snapshot 1`] = `
preact-render-spy (1 nodes)
-------
<div
  class="ada-embed-intro-blurb ada-embed-intro-blurb--show"
  style="right: 104px;"
>
  <button
    class="ada-embed-intro-blurb__dismiss-button"
    title="Dismiss Intro"
    onClick={[Function bound ]}
  >
  </button>
  <p
    class="ada-embed-intro-blurb__message"
    aria-live="assertive"
    role="alert"
    onClick={[Function bound ]}
    onKeyPress={[Function bound ]}
    tabIndex={0}
  >
  </p>
</div>

`;
